### PR TITLE
Websocket stuff

### PR DESCRIPTION
**Pull request by pubrrr**

### CODE REVIEW

File: client/page/index.html [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    socket.send(b);
+    socket.send(' ');
+    socket.send('blabla');
+
+    app.ports.sendMessage.subscribe(
+        message => {
+            console.log(message);
+            socket.send(message)
+        }
```
Comment (ekinimo) [inline client/page/index.html:33]: do we need js code or is it just for testing?

File: client/src/Main.elm [hunk 2]
```diff
@@ -1,5 +1,56 @@
...
+        , subscriptions = subscriptions
+        }
+
+
+type Msg
+    = Increment
+    | Decrement
+
+
+type alias Model =
```
Comment (ekinimo) [inline client/src/Main.elm:26]: that is the magic type that encapsulate the state right

File: client/src/Main.elm [hunk 3]
```diff
@@ -1,5 +1,56 @@
...
+import Json.Encode exposing (Value, string)
 
-main = text "hi"
\ No newline at end of file
+
+port sendMessage : Value -> Cmd msg
+
+
+main =
+    Browser.element
```
Comment (ekinimo) [inline client/src/Main.elm:13]: what does this  element function do ?

File: client/src/Main.elm [hunk 4]
```diff
@@ -1,5 +1,56 @@
...
+        , button [ onClick Increment ] [ text "+" ]
+        ]
+
+
+init : () -> ( Model, Cmd Msg )
+init _ =
+    ( 0, Cmd.none )
+
+
+subscriptions : Model -> Sub Msg
```
Comment (ekinimo) [inline client/src/Main.elm:54]: why do we need this ?

File: server/src/main.rs [hunk 5]
```diff
@@ -1,3 +1,153 @@
...
+use std::sync::mpsc::channel;
+use std::thread::{sleep, spawn};
+use std::time::Duration;
+
+use http::header::{CONNECTION, SEC_WEBSOCKET_ACCEPT, UPGRADE};
+use http::{HeaderMap, Response, Version};
+use httparse::{Request, EMPTY_HEADER};
+use sha1_smol::Sha1;
+
 fn main() {
```
Comment (ekinimo) [inline server/src/main.rs:7]: this is really confusing

File: server/src/main.rs [hunk 6]
```diff
@@ -1,3 +1,153 @@
...
-    println!("Hello, world!");
+    let server = TcpListener::bind("127.0.0.1:8080").unwrap();
+    server.set_nonblocking(true).unwrap();
+
+    let (sender, receiver) = channel();
+
+    let mut clients = vec![];
+
+    loop {
```
Comment (ekinimo) [inline server/src/main.rs]: there must be a better way than infintely looping and waiting for server connections some kind of a event handler some reactive framework etc.

**Review by ekinimo — APPROVED** [review]

File: client/src/Main.elm (hunk 2 above)
Comment (pubrrr) [reply client/src/Main.elm:26]: Yes

File: client/src/Main.elm (hunk 3 above)
Comment (pubrrr) [reply client/src/Main.elm:13]: It's some kind of convenience function to create a `Programm`.

File: client/src/Main.elm (hunk 4 above)
Comment (pubrrr) [reply client/src/Main.elm:54]: Probably not yet. Later to listen for messages from the port.

File: server/src/main.rs (hunk 6 above)
Comment (pubrrr) [reply server/src/main.rs]: The framework will probably do exactly the same. But if there is a good framework, we might try it.

File: client/page/index.html (hunk 1 above)
Comment (pubrrr) [reply client/page/index.html:33]: the console.log yes, the socket.send is the crucial part.

**Pull request merged**